aiida/cmdline/commands/cmd_code.py

```diff
@@ -296,117 +297,103 @@ def relabel(code, label):
         echo.echo_success(f'Code<{code.pk}> relabeled from {old_label} to {code.full_label}')
 
 
+VALID_CODE_PROJECTIONS = (
+    'full_label',
+    'entry_point',
+    'pk',
+    'default_calc_job_plugin',
+    'computer',
+    'user',
+)
+
+
 @verdi_code.command('list')
 @options.COMPUTER(help='Filter codes by computer.')
 @options.INPUT_PLUGIN(help='Filter codes by calculation input plugin.')
 @options.ALL(help='Include hidden codes.')
 @options.ALL_USERS(help='Include codes from all users.')
+@options.PROJECT(type=click.Choice(VALID_CODE_PROJECTIONS), default=['full_label', 'pk', 'entry_point'])
 @click.option('-o', '--show-owner', 'show_owner', is_flag=True, default=False, help='Show owners of codes.')
 @with_dbenv()
-def code_list(computer, input_plugin, all_entries, all_users, show_owner):
+def code_list(computer, input_plugin, all_entries, all_users, show_owner, project):
+    # pylint: disable=too-many-locals,too-many-branches
     """List the available codes."""
     from aiida import orm
-    from aiida.orm import Code  # pylint: disable=redefined-outer-name
+    from aiida.orm.utils.node import load_node_class
+
+    if show_owner:
+        echo.echo_deprecated(
+            'the `-o/--show-owner` option is deprecated. To show the user use the `-P/--project` option instead, e.g., '
+            '`verdi code list -P full_label user`.'
+        )
+        if 'user' not in project:
+            project = project + ('user',)
+
+    project_mapping = {
+        'full_label': [('code', 'label'), ('computer', 'label')],
+        'entry_point': [('code', 'node_type')],
+        'pk': [('code', 'id')],
+        'default_calc_job_plugin': [('code', 'attributes.default_calc_job_plugin')],
+        'computer': [('computer', 'label')],
+        'user': [('user', 'email')],
+    }
+
+    filters = defaultdict(dict)
+    projections = defaultdict(list)
+
+    for key in project:
+        for entity, projection in project_mapping[key]:
+            if projection not in projections[entity]:
+                projections[entity].append(projection)
+
+    if not all_entries:
+        filters['code'][f'extras.{orm.Code.HIDDEN_KEY}'] = {'!==': True}
 
-    qb_user_filters = {}
     if not all_users:
-        user = orm.User.collection.get_default()
-        qb_user_filters['email'] = user.email
+        filters['user']['email'] = orm.User.collection.get_default().email
 
-    qb_computer_filters = {}
     if computer is not None:
-        qb_computer_filters['label'] = computer.label
+        filters['computer']['uuid'] = computer.uuid
 
-    qb_code_filters = {}
     if input_plugin is not None:
-        qb_code_filters['attributes.input_plugin'] = input_plugin.name
+        filters['code']['attributes.input_plugin'] = input_plugin.name
+
+    echo.echo_report('Use `verdi code show IDENTIFIER` to see details for a code')
```

Comment:
I would generally suggest putting instructions like these at the end of an output, since at the top they can be hidden when the list gets long.
Perhaps also lose the `Report:` prefix.